### ISSUE TITLE
Prevent access to internal properties from Ecmascript code even with buffer access

### ISSUE THREAD

**Issue by svaarala** [svaarala/duktape#44]
The current internal property approach is not ideal for sandboxing untrusted code. The sandbox must prevent user code from creating or using buffer values, because a buffer value can be used to construct a non-UTF-8 string which can then be used to access internal properties of objects. Depending on the writability of the internal properties, this may or may not be a sandboxing security issue.

This limitation is OK for sandboxing standard Ecmascript code which needs no access to buffer values at all. But there are also sandboxed environments where buffer values need to be available.

Not sure what's the best fix for this. A few possibilities:
- Deny access to internal properties from Ecmascript code even if the calling code uses an internal property name. Internal properties would be non-readable and non-writable (e.g. attempt to read would return `undefined` and attempt to write would throw an error).
- Deny access to internal properties from both C and Ecmascript code, unless accessed with an API call intended specifically for internal properties (equivalently, use a flag for this purpose). This would prevent intentional or accidental access to internal properties both from C and Ecmascript code.
- Deny buffer-to-string coercion which would result in a non-extended-UTF-8 string. This would prevent Ecmascript code from constructing internal property names (strings), but wouldn't prevent access if an internal property name was constructed by C code and leaked to Ecmascript code.

Somewhat unrelated, an alternative to using a `\xFF` prefix would be to add a new non-standard "hidden" attribute to object properties. A hidden property would behave like an internal property does now, but would not need to have a special form. This cannot be implemented as a `duk_hstring` flag because a certain property name may be a hidden property in one object but not in another.

**Comment by svaarala** [comment]
Internal properties should also be aligned with ES6 Symbols:

- https://curiosity-driven.org/private-properties-in-javascript

**Comment by svaarala** [comment]
Prototype branch with a draft that hides internal properties from Ecmascript code but allows user C code to access them:

- https://github.com/svaarala/duktape/compare/hide-internal-properties

This should suffice for sandboxing, but:

- The branch adds an extra flags field to property methods which is of course additional overhead (quite useful for other things too, though).
- It's easy for a user C function to bypass the sandboxing by doing a property lookup using an Ecmascript-supplied key. So it might be better to have an entirely separate C API call to access an internal property (perhaps aligned to the ES6 symbols, e.g. `duk_get_prop_symbol()` or similar).

**Comment by fatcerberus** [comment]
I suppose this means Duktape 2.0 will include support for Symbols as well? :)

**Comment by svaarala** [comment]
Not necessarily, but if changes are made to the internal property model, there should at least be a plan how symbols will be implemented without needing another incompatible change :)

**Comment by svaarala** [comment]
I'll move this to 3.0.0 because the symbol support is still experimental and making this change makes more sense when that behavior is finalized.

**Comment by fatcerberus** [comment]
The only Ecmascript binding remaining that can create hidden symbols is `Duktape.dec()` right?  So removing the `Duktape` binding should be enough of a stopgap until this is implemented.

**Comment by svaarala** [comment]
Actually no, `Duktape.dec()` creates a plain buffer as its output right now (maybe changing to Uint8Array to use standard types); one could create a similar buffer by other means. When you convert that buffer to a string, an encoding is applied just like with any other buffer.

**Comment by svaarala** [comment]
So the remaining risk is C code that pushes unvalidated / un-encoded data using `duk_push_string()` and creates a hidden Symbol by accident. That's avoided cleanly if C code always encodes/validates strings it pushes to be CESU-8; any other formats will create weirdly behaving strings in any case.
